### PR TITLE
fix (net5): handle `RejectedExecutionException` happening for ubuntu IO completed operations

### PR DESCRIPTION
**Pull request by DeagleGross**
There were 2 problems with the current code (on ubuntu):
1) (change in `src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.cs`) Test host was consistently crashing due to `RejectedExecutionException`. The problem is that closing the socket (invoking `_clientChannel.CloseAsync()`) eventually calls [SocketAsyncContext.TryCancel()](https://github.com/dotnet/runtime/blob/c315bac8ae7cfc613cb446ca594a17d07d02aa3d/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L264C21-L264C125) which results in another IO callback. 

When DotNetty handles the IO callback, eventLoop can be already in the `Shutdown` state, and that throws the `RejectedExecutionException`. However, we can try to skip scheduling the work on the eventLoop, if channel (actual socket) is already closed. We can also validate that it is a socket closure operation raised from `System.Net.Sockets.SocketAsyncEventArgs.TryCancel()` by checking the `operation.SocketError` if it equals `SocketError.OperationAborted`. 

2) (change in `src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.Unsafe.cs`) I confirmed that some of tests fail due to not sending the HTTP2 Preface data on the connection. The reason is - not always preface sending called from `HandlerAdded(ctx)` API is reliable, because channel can be not active at a time. And we need to repeat the send invoking it from `ChannelActive(ctx)` event. However, I saw that `channel.ConnectPromise` is null and if it is, then we never invoke `ChannelActive` (however we should). So I changed the code to invoke it even if connectPromise is null (not changing the code, but only the order).

Other tests rarely fail, but I can see they being flaky. We can deal with those later

### CODE REVIEW

File: test/DotNetty.Transport.Libuv.Tests/EventLoopTests.cs [hunk 1]
```diff
@@ -82,7 +82,7 @@ public void ShutdownAfterExecute()
             Assert.Throws<RejectedExecutionException>(() => this.eventLoop.Execute(this.noOp));
         }
 
-        [Fact]
+        [Fact(Skip = "flaky")]
```
Comment (tomkerkhove) [inline test/DotNetty.Transport.Libuv.Tests/EventLoopTests.cs:85]: Do we have an issue/bug to track this?

**Review by tomkerkhove — APPROVED** [review]

**Comment by maksimkim** [comment]
Do we need all the extra logging code? I assume a lot of it can be covered by existing LoggingHandler

File: src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.Unsafe.cs [hunk 2]
```diff
@@ -64,20 +68,21 @@ public sealed override Task ConnectAsync(EndPoint remoteAddress, EndPoint localA
...
                 try
                 {
-                    if (ch._connectPromise is object)
+                    if (ch._connectPromise is not null)
                     {
                         ThrowHelper.ThrowInvalidOperationException_ConnAttemptAlreadyMade();
                     }
 
                     bool wasActive = _channel.IsActive;
+                    var connectPromise = ch.NewPromise(remoteAddress);
```
Comment (maksimkim) [inline src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.Unsafe.cs]: do we need to initialize _connectPromise field in case DoConnect returns true?

File: src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.cs [hunk 3]
```diff
@@ -235,6 +234,15 @@ static void OnIoCompleted(object sender, SocketAsyncEventArgs args)
             var channel = operation.Channel;
             var @unsafe = channel.Unsafe;
             IEventLoop eventLoop = channel.EventLoop;
+            
+            if (operation.SocketError == SocketError.OperationAborted // means System.Net.Sockets.Socket was closed. Most probably we received a callback for closure, no for real IO happened
```
Comment (maksimkim) [inline src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.cs]: > Most probably the callback was invoked because socket was closed, not because IO operation completed.

File: test/DotNetty.Codecs.Http2.Tests/Http2ConnectionRoundtripTest.cs [hunk 4]
```diff
@@ -97,17 +97,15 @@ protected override void SetupBootstrap(Bootstrap bootstrap)
             bootstrap.Group(new DefaultEventLoopGroup()).Channel<LocalChannel>();
         }
 
-        protected override void StartBootstrap()
+        protected override async Task StartBootstrap()
```
Comment (maksimkim) [inline test/DotNetty.Codecs.Http2.Tests/Http2ConnectionRoundtripTest.cs:100]: what is the purpose of this change?

File: src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.Unsafe.cs (hunk 2 above)
Comment (DeagleGross) [reply src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.Unsafe.cs]: changed

File: src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.cs (hunk 3 above)
Comment (DeagleGross) [reply src/DotNetty.Transport/Channels/Sockets/AbstractSocketChannel.cs]: wrote it down

File: test/DotNetty.Codecs.Http2.Tests/Http2ConnectionRoundtripTest.cs (hunk 4 above)
Comment (DeagleGross) [reply test/DotNetty.Codecs.Http2.Tests/Http2ConnectionRoundtripTest.cs:100]: Some Codecs.Http2.Tests are very similar in terms of preparation for the test, but are re-written manually a bit differently (in example `void` in one test and `Task` in another). Moreover I needed the logs to see if the actual ConnectAsync operation completed on the client or server channel, and it took me multiple tries to emit logs for different classes.

Because of this poor copy-paste code, I introduced a single class `HttpToHttp2ConnectionHandlerTest` which has the API which also logs the operations. That is why I needed to change other places to be in sync, and made this change from `void` to `Task`.

Why do u feel bad about this?

**Comment by DeagleGross** [comment]
> Do we need all the extra logging code? I assume a lot of it can be covered by existing LoggingHandler

@maksimkim , I could not enable it for the tests (when I specified `verbosity=detailed`). Please, clarify on how this LoggingHandler should be enabled and logs can be seen, because that is not scalable to add logs, see stack traces, make changes, and then delete logging every time. I think we should have it enabled if specified (or in Debug mode) to have the visibility into the operations.

Also, I think putting logs only in Debug mode via directive `#if` is not impacting the production logging anyhow. Only for the inner library visibility in debug mode (tests)

**Comment by maksimkim** [comment]
LoggingHandler is enabled by adding it to the channel pipeline. Once added it starts logging all events in the pipeline.

If we really need extra logging on top of that, I don't think "#if DEBUG" is required. Putting it under debug level logging should be enough.

**Comment by DeagleGross** [comment]
Makes sense, @maksimkim .
I rollbacked the logging I added, I will take a look on how `LoggingHandler` is set in the tests, and will try to onboard missing logs emitted for debugging during next changes.

File: test/DotNetty.Transport.Libuv.Tests/EventLoopTests.cs (hunk 1 above)
Comment (DeagleGross) [reply test/DotNetty.Transport.Libuv.Tests/EventLoopTests.cs:85]: [Bug 29658901](https://msazure.visualstudio.com/One/_workitems/edit/29658901): Fix flaky test/DotNetty.Transport.Libuv.Tests/EventLoopTests.ScheduleTask

File: src/DotNetty.Transport/Internal/LoggingExtensions.cs [hunk 5]
```diff
@@ -26,11 +26,19 @@
...
 namespace DotNetty.Transport
 {
     internal static class TransportLoggingExtensions
-    {
+    {       
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static void AbstractSocketIoCallbackSkipped<TChannel, TUnsafe>(this IInternalLogger logger, SocketChannelAsyncOperation<TChannel,TUnsafe> operation, TChannel channel, IEventLoop eventLoop) 
+            where TChannel : AbstractSocketChannel<TChannel, TUnsafe> where TUnsafe : AbstractSocketChannel<TChannel, TUnsafe>.AbstractSocketUnsafe, new()
+        {
+            logger.Debug($"Skipping callback schedule due to eventLoop closure and socket already closed. operation: {operation.LastOperation} / socketError: {operation.SocketError}; channel: {channel.Id} (IsOpen: {channel.IsOpen})");
```
Comment (maksimkim) [inline src/DotNetty.Transport/Internal/LoggingExtensions.cs]: the eventLoop part of the log message is irrelevant.

File: test/DotNetty.Codecs.Http2.Tests/Http2ConnectionRoundtripTest.cs (hunk 4 above)
Comment (maksimkim) [reply test/DotNetty.Codecs.Http2.Tests/Http2ConnectionRoundtripTest.cs:100]: I see that in 2 our of 3 invocations of StartBootstrap it is awaited synchronously anyway, so I wonder why it can't be void. But given xunit supports async tests, eventually we would want to refactor it from .GetResults() to awaits anyway, so this change is fine.

File: src/DotNetty.Transport/Internal/LoggingExtensions.cs (hunk 5 above)
Comment (DeagleGross) [reply src/DotNetty.Transport/Internal/LoggingExtensions.cs]: removed eventLoop as parameter

Comment (DeagleGross) [reply src/DotNetty.Transport/Internal/LoggingExtensions.cs]: and changed the log message

**Review by maksimkim — APPROVED** [review]
